bls/src/main/java/tech/pegasys/artemis/bls/mikuli/BLS12381.java

```diff
@@ -42,6 +45,43 @@
  */
 public final class BLS12381 {
 
+  private static final long MAX_BATCH_VERIFY_RANDOM_MULTIPLIER = Long.MAX_VALUE;
+
+  private static Random getRND() {
+    // Milagro RAND has some issues with generating 'small' random numbers
+    // and is not thread safe
+    // Using non-secure random due to the JDK Linux secure random issue:
+    // https://bugs.java.com/bugdatabase/view_bug.do?bug_id=6521844
+    // A potential attack here has a very limited application and is not feasible
+    // Thus using non-secure random doesn't significantly mitigate the security
+    return ThreadLocalRandom.current();
```

Comment:
That linux bug was fixed in JDK 7 which we don't support anymore.  Java SecureRandom now uses `/dev/urandom`.

Reply:
Yeah, mixed up `Updated` and `Resolved` dates. Though it just was recently fixed. 
However while running the test with `SecureRandom` on Linux (I reproduced on Ubuntu 18, the same behavior was on the Travis) it stuck for 30 mins with this stack trace: 
```
    java.lang.Thread.State: RUNNABLE
        at java.io.FileInputStream.readBytes(java.base@11.0.7/Native Method)
        at java.io.FileInputStream.read(java.base@11.0.7/FileInputStream.java:279)
        at java.io.FilterInputStream.read(java.base@11.0.7/FilterInputStream.java:133)
        at sun.security.provider.NativePRNG$RandomIO.readFully(java.base@11.0.7/NativePRNG.java:424)
        at sun.security.provider.NativePRNG$RandomIO.ensureBufferValid(java.base@11.0.7/NativePRNG.java:526)
        at sun.security.provider.NativePRNG$RandomIO.implNextBytes(java.base@11.0.7/NativePRNG.java:545)
        - locked <0x00000007004ac3c0> (a java.lang.Object)
        at sun.security.provider.NativePRNG$Blocking.engineNextBytes(java.base@11.0.7/NativePRNG.java:268)
        at java.security.SecureRandom.nextBytes(java.base@11.0.7/SecureRandom.java:742)
        at java.security.SecureRandom.next(java.base@11.0.7/SecureRandom.java:799)
        at java.util.Random.nextLong(java.base@11.0.7/Random.java:424)
        at tech.pegasys.artemis.bls.mikuli.BLS12381.nextBatchRandomMultiplier(BLS12381.java:325)
        at tech.pegasys.artemis.bls.mikuli.BLS12381.prepareBatchVerify(BLS12381.java:240)
        at tech.pegasys.artemis.bls.mikuli.BLS12381Test.lambda$batchVerifyTest$4(BLS12381Test.java:178)
        at tech.pegasys.artemis.bls.mikuli.BLS12381Test$$Lambda$464/0x00000008401ec040.apply(Unknown Source)
```

Here are opened files: 
```
ubuntu@ip-172-31-30-114:~/jvm-libp2p$ lsof -p 16745 | grep random
java    16745 ubuntu   96r      CHR                1,8       0t0     10 /dev/random
java    16745 ubuntu   97r      CHR                1,9       0t0     11 /dev/urandom
```

Also found another related JDK bug (which was also fixed) : https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8098581

As we discussed with @benjaminion this place is not critical for using _secure_ random, but there is `SecureRandomProvider` which may potentially hit the same issue. I will add an issue to the Jira

Reply:
Here is the java version: 
```
openjdk version "11.0.7" 2020-04-14
OpenJDK Runtime Environment (build 11.0.7+10-post-Ubuntu-2ubuntu218.04)
OpenJDK 64-Bit Server VM (build 11.0.7+10-post-Ubuntu-2ubuntu218.04, mixed mode, sharing)
```
Here is the commit on which the issue is reproduced: https://github.com/PegaSysEng/teku/pull/1632/commits/c3e8c2e34c5af675364f3f771f868209543d5d01

Reply:
hmm, we went through this in a lot of detail as part of Pantheon's security audit and confirmed Java would use `/dev/urandom`.  Originally Pantheon had overridden it to use `/dev/random` because that's "more secure" (the auditors disagreed, arguments ensued, `/dev/urandom` eventually won because it made the auditors happy and actually worked).

It looks a bit more complex than I remember it and is now trying to look up egdSource via security properties - specifically the `securerandom.source` security property. Interestingly AdoptOpenJDK 11 on Mac sets that to file:/dev/random so potentially Java has changed again.  I'm a little surprised Besu hasn't run into this as it uses a *lot* of entropy (mostly from discovery messages IIRC).

I'm personally not too concerned about using plain `Random` here though I suspect we'll get dinged for it in the audit.  And we're going to have to ensure people at least know how to select /dev/urandom as the source instead of /dev/random and possibly find a way to set it as the default.